design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,207 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a custer-level rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object without interupting the workload.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to eventually propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out ReplicaSets.
+
+
+## Goals
+
+-   Provide a way to eventually propagate changes from a running VM object to its corresponding VMI without interrupting the workload.
+    
+-   Maintain backwards compatibility for cluster owners that prefer keeping staged changes until next VM reboot.
+    
+
+## Non Goals
+
+-   Live uppdate for **all** fields of the Virtual Machine API. Only select fields can be propagated to the guest. 
+-   Make LiveUpdate the default. We defer this decision.
+-   Support upgrades of VMs that are using the `liveUpdateFeatures`.
+-   This design does consider the hot-plugging of network interfaces and additional storage.
+
+## Definition of Users
+
+A VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a VM owner I want changes that I applied to the VM object to propagate to the VMI if possible.
+
+As a VM owner, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+> **Note:** Below we are only discussing changes to fields that support live updates.
+                  References to `all` fields only refers to fields that can be live-updated. 
+
+## KubeVirt CR
+
+A new field called `vmRolloutStrategy` will be introduced in the KubeVirt CR. This field will control the behavior of the VM objects. Two possible values will be available:
+
+1. `LiveUpdate` means all changes to the VM object will be eventually applied to the VMI.
+    
+2. `Stage` would mean the current familiar behavior remains - all changes to the VM object will be staged and applied after a restart.
+
+In both cases, a condition called `RestartRequired` will be added to the VM to reflect the need for a restart and the reason why.
+
+The default value of the `vmRolloutStrategy` field if not specified will be `Stage`.
+
+The `vmRolloutStrategy` field will be usable only if the `VMLiveUpdateFeatures` feature gate is enabled.
+
+
+## VM API
+
+The VM level `liveUpdateFeatures` field will be removed.
```

Comment:
What happens if VMs are already using this (due to using some TP features) - would we silently remove that/ignore it? Or would we issue a warning that the VM definition needs to be changed?

Reply:
The struct would be removed from the API definition, VMs with this struct will either be silently accepted (and the struct dropped) or rejected. @jean-edouard how will the system behave if we remove the api?

Reply:
This is a good question: what happens to existing k8s objects when an update to their API removes fields used by said objects?
I'll try and figure out a way to test this locally and let you know the outcome.
Either way though, this is just not a supported use-case.

Reply:
We appear to be lucky on this one. I deployed a cluster with KubeVirt 1.1.0 and a VM with some liveUpdateFeatures set.
Then I updated to a custom version with liveUpdateFeatures removed (but no code to deal with existing VMs).
I then looked at the VM yaml and found that the liveUpdateFeatures block was just gone.

Reply:
k8s is usually not removing fields, they are just left in place with a deprecation notice and a no-op.

See https://github.com/kubevirt/community/pull/251/files#diff-845f63068ea8ebf517e69fa5547ab723ec6279b6743bbc15b9a2de895c9bdbf0R241

Reply:
It is an issue with the DB in the api-server, the ref I have there has all the discussion about it. k8s usually do not ever remove structs even if in Alpha, they just mark them as deprecated and they are in practice a no-op with a build and possibly runtime warning.

I think the topic itself is a bit unclear, I for sure do not understand fully their reasoning and therefore it is hard to be sure if ignoring it is possible or not.

Reply:
Okay let me reprhase:
API support wise these fields will be removed.
The technical implementation of this has to be seen,